### PR TITLE
Porting code from PyTorch 0.3.1 to PyTorch 1.8.1

### PR DESCRIPTION
**Pull request by animesh-007**
I have ported the code to the latest version. Can you review it?

### CODE REVIEW

File: classifiers/classifier_images.py [hunk 1]
```diff
@@ -71,11 +71,11 @@ def fit_zsl(self):
                    
                 # inputv = Variable(self.input)
                 # labelv = Variable(self.label)
-                inputv = self.input.clone().detach()
-                labelv = self.label.clone().detach()
+                inputv = self.input.detach()
```
Comment (adityac8) [inline classifiers/classifier_images.py]: Check if detach is required.

File: classifiers/classifier_images.py [hunk 2]
```diff
@@ -71,8 +71,8 @@ def fit_zsl(self):
                    
                 # inputv = Variable(self.input)
                 # labelv = Variable(self.label)
-                inputv = self.input.detach()
-                labelv = self.label.detach()
+                inputv = self.input.clone()
```
Comment (adityac8) [inline classifiers/classifier_images.py]: Check if `clone` is required. It might also work without using clone. If it works, we might need to remove some redundant variables.

File: classifiers/classifier_images.py [hunk 3]
```diff
@@ -162,15 +162,16 @@ def val_gzsl(self, test_X, test_label, target_classes):
...
-                inputX = test_X[start:end].detach().cuda()
-            else:
-                # inputX = Variable(test_X[start:end], volatile=True)
-                inputX = test_X[start:end].detach()
-            output = self.model(inputX)  
-            _, predicted_label[start:end] = torch.max(output.data, 1)
-            start = end
+            with torch.no_grad():
+                if self.cuda:
+                    # inputX = Variable(test_X[start:end].cuda(), volatile=True)
```
Comment (adityac8) [inline classifiers/classifier_images.py]: During testing, make sure you use the following correct code.
```
model.eval()
with torch.no_grad():
    pred = model(data)
model.train()
```

File: classifiers/classifier_images.py (hunk 2 above)
Comment (animesh-007) [reply classifiers/classifier_images.py]: Okay.

File: classifiers/classifier_images.py [hunk 4]
```diff
@@ -190,15 +191,16 @@ def val(self, test_X, test_label, target_classes):
...
-                inputX = test_X[start:end].detach().cuda()
-            else:
-                # inputX = Variable(test_X[start:end], volatile=True)
-                inputX = test_X[start:end].detach()
-            output = self.model(inputX) 
-            _, predicted_label[start:end] = torch.max(output.data, 1)
-            start = end
+            with torch.no_grad():
+                if self.cuda:
+                    # inputX = Variable(test_X[start:end].cuda(), volatile=True)
```
Comment (adityac8) [inline classifiers/classifier_images.py]: Define the variable 
```
inputX = test_X[start:end].clone()
if cuda:
    inputX = inputX.cuda()
```

File: classifiers/classifier_images.py (hunk 3 above)
Comment (animesh-007) [reply classifiers/classifier_images.py]: Okay. I will check it again.

File: train_images.py [hunk 5]
```diff
@@ -308,19 +308,20 @@ def calc_gradient_penalty(netD,real_data, fake_data, input_att):
...
+    with torch.no_grad():
+        syn_feature, syn_label = generate_syn_feature(netG,data.unseenclasses, data.attribute, opt["network"]["gan"]["syn_num"],netF=netF,netDec=netDec)
+        # Generalized zero-shot learning
+        if opt["network"]["classifier"]["gzsl"]:   
+            # Concatenate real seen features with synthesized unseen features
+            train_X = torch.cat((data.train_feature, syn_feature), 0)
+            train_Y = torch.cat((data.train_label, syn_label), 0)
+            nclass = opt["network"]["gan"]["num_class"]
+            # Train GZSL classifier
+            gzsl_cls = classifier.CLASSIFIER(train_X, train_Y, data, nclass, cuda, opt["network"]["classifier"]["lr"], 0.5, \
```
Comment (animesh-007) [inline train_images.py]: To-Do: Fixing this.

File: classifiers/classifier_images.py [hunk 6]
```diff
@@ -102,8 +102,8 @@ def fit(self):
                    
                 # inputv = Variable(self.input)
                 # labelv = Variable(self.label)
-                inputv = self.input.clone()
-                labelv = self.label.clone()
+                inputv = self.input
```
Comment (adityac8) [inline classifiers/classifier_images.py:128]: Remove redundant variable

File: classifiers/classifier_images.py [hunk 7]
```diff
@@ -162,17 +162,18 @@ def val_gzsl(self, test_X, test_label, target_classes):
...
             end = min(ntest, start+self.batch_size)
+            inputX = test_X[start:end]
             with torch.no_grad():
                 if self.cuda:
                     # inputX = Variable(test_X[start:end].cuda(), volatile=True)
-                    inputX = test_X[start:end].clone().cuda()
-                else:
-                    # inputX = Variable(test_X[start:end], volatile=True)
-                    inputX = test_X[start:end].clone()
+                    inputX.cuda()
```
Comment (adityac8) [inline classifiers/classifier_images.py]: inputX = inputX.cuda()

File: train_images.py [hunk 8]
```diff
@@ -248,14 +248,14 @@ def calc_gradient_penalty(netD,real_data, fake_data, input_att):
...
-            input_attv = input_att.clone()
+            input_resv = input_res
+            input_attv = input_att
             means, log_var = netE(input_resv, input_attv)
             std = torch.exp(0.5 * log_var)
             eps = torch.randn([opt["train"]["batch_size"], opt["network"]["gan"]["latent_dim"]]).cpu()
 
             # eps = Variable(eps.cuda())
-            eps = eps.clone().cuda()
+            eps = eps.cuda()
```
Comment (adityac8) [inline train_images.py]: why no if cuda

File: train_images.py [hunk 9]
```diff
@@ -310,18 +310,21 @@ def calc_gradient_penalty(netD,real_data, fake_data, input_att):
...
-            train_X = torch.cat((data.train_feature, syn_feature), 0)
-            train_Y = torch.cat((data.train_label, syn_label), 0)
-            nclass = opt["network"]["gan"]["num_class"]
-            # Train GZSL classifier
-            gzsl_cls = classifier.CLASSIFIER(train_X, train_Y, data, nclass, cuda, opt["network"]["classifier"]["lr"], 0.5, \
-                    25, opt["network"]["gan"]["syn_num"], generalized=True, netDec=netDec, dec_size=opt["network"]["gan"]["att_size"], dec_hidden_size=4096)
-            if best_gzsl_acc < gzsl_cls.H:
-                best_acc_seen, best_acc_unseen, best_gzsl_acc = gzsl_cls.acc_seen, gzsl_cls.acc_unseen, gzsl_cls.H
-            logger.info('GZSL: seen=%.4f, unseen=%.4f, h=%.4f' % (gzsl_cls.acc_seen, gzsl_cls.acc_unseen, gzsl_cls.H))
+    #  @adityac8 netG.train() is not defined by authour here.
```
Comment (adityac8) [inline train_images.py]: Check https://github.com/channelCS/ZSL_Generative/pull/3/files/d15ddbdf3a0a3f4273b11b48686ce4d7c58c335a..f77904ebbf9ff27e4c31f589a565042f7f8ea8f1#r647485770

File: classifiers/classifier_images.py [hunk 10]
```diff
@@ -61,13 +57,15 @@ def fit_zsl(self):
...
         last_loss_epoch = 1e8 
+        self.model.train()
         best_model = copy.deepcopy(self.model.state_dict())
         for epoch in range(self.nepoch):
             for i in range(0, self.ntrain, self.batch_size):      
                 self.model.zero_grad()
                 batch_input, batch_label = self.next_batch(self.batch_size) 
                 self.input.copy_(batch_input)
                 self.label.copy_(batch_label)
+                self.optimizer.zero_grad()
```
Comment (adityac8) [inline classifiers/classifier_images.py]: This is wrong
